src/bundle.ts

```diff
@@ -0,0 +1,91 @@
+import * as path from "path";
+import { TranspiledFile, EmitHost } from "./Transpile";
+import { SourceNode } from "source-map";
+import { formatPathToLuaPath, trimExtension, normalizeSlashes } from "./utils";
+import { Diagnostic, Program } from "typescript";
+import { couldNotFindBundleEntryPoint } from "./diagnostics";
+
+const formatPath = (path: string) => formatPathToLuaPath(trimExtension(path));
+
+export function bundleTranspiledFiles(
+    bundleFile: string,
+    entryModule: string,
+    transpiledFiles: TranspiledFile[],
+    program: Program,
+    emitHost: EmitHost
+): [Diagnostic[], TranspiledFile] {
+    const diagnostics: Diagnostic[] = [];
+
+    const resolvedEntryModule = resolveAbsolutePath(program, entryModule);
+    if (!transpiledFiles.some(f => normalizeSlashes(f.fileName) === normalizeSlashes(resolvedEntryModule))) {
+        return [[couldNotFindBundleEntryPoint(entryModule)], { fileName: bundleFile }];
+    }
+
+    // For each file: ["<file name>"] = function() <lua content> end,
+    const projectRootDir = getProjectRootDir(program);
+    const moduleTableEntries: SourceChunk[] = transpiledFiles.map(f => moduleSourceNode(f, projectRootDir));
+
+    // If any of the modules contains a require for lualib_bundle, add it to the module table.
+    const lualibRequired = transpiledFiles.some(f => f.lua && f.lua.match(/require\("lualib_bundle"\)/));
+    if (lualibRequired) {
+        const lualibBundle = emitHost.readFile(path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua"));
+        moduleTableEntries.push(`["lualib_bundle"] = function() ${lualibBundle} end,\n`);
+    }
+
+    // Create ____modules table containing all entries from moduleTableEntries
+    const moduleTable = createModuleTableNode(moduleTableEntries);
+
+    // Override `require` to read from ____modules table.
+    const requireOverride =
+        `local ____originalRequire = require\n` +
+        `function require(file) if ____modules[file] then return ____modules[file]() ` +
+        `else print("Could not find module '"..file.."' to require."); return ____originalRequire(file) end end\n`;
```

Comment:
```suggestion
        `else return ____originalRequire(file) end end\n`;
```

Reply:
Not sure if this should be removed. Imagine an environment where require is nil (so we need bundling), you will get an 'attempted to call nil value' error, which is really unclear.

Reply:
Well, if that's a problem we just issue an error when `____originalRequire` is nil. Falling back to original require is important when there are environment-provided modules, and we shouldn't do any IO in these cases